src/compass/geo_cslc.py

```diff
@@ -0,0 +1,196 @@
+import os
+import time
+
+import isce3
+import journal
+import numpy as np
+from osgeo import gdal
+from shapely.geometry import MultiPoint
+
+from compass.utils.geo_runconfig import GeoRunConfig
+from compass.utils.range_split_spectrum import range_split_spectrum
+from compass.utils.yaml_argparse import YamlArgparse
+
+
+def run(cfg):
+    '''
+    Run geocode burst workflow with user-defined
+    args stored in dictionary runconfig *cfg*
+
+    Parameter:
+    ---------
+    cfg: dict
+       Dictionary with user runconfig options
+    '''
+    info_channel = journal.info("geo_cslc.run")
+
+    # Start tracking processing time
+    t_start = time.time()
+    info_channel.log("Starting geocode burst")
+
+    # Common initializations
+    dem_raster = isce3.io.Raster(cfg.dem)
+    epsg = dem_raster.get_epsg()
+    proj = isce3.core.make_projection(epsg)
+    ellipsoid = proj.ellipsoid
+    image_grid_doppler = isce3.core.LUT2d()
+    threshold = cfg.geo2rdr_params.threshold
+    iters = cfg.geo2rdr_params.numiter
+    blocksize = cfg.geo2rdr_params.lines_per_block
+    dem_margin = cfg.geocoding_params.dem_margin
+    flatten = cfg.geocoding_params.flatten
+
+    for bursts in cfg.bursts:
+        burst_id = bursts[0].burst_id
+        geo_grid = cfg.geogrids[burst_id]
+
+        output_path = f'{cfg.product_path}/{burst_id}'
+        scratch_path = f'{cfg.scratch_path}/{burst_id}'
+        os.makedirs(output_path, exist_ok=True)
+        os.makedirs(scratch_path, exist_ok=True)
+        for burst in bursts:
+            radar_grid = burst.as_isce3_radargrid()
+            native_doppler = burst.doppler.lut2d
+            orbit = burst.orbit
+
+            # Get azimuth polynomial coefficients for this burst
+            az_carrier_poly2d = burst.get_az_carrier_poly()
+
+            # Split the range bandwidth of the burst, if required
+            if cfg.split_spectrum_params.enabled:
```

Comment:
I noticed that I had a bug in the split-spectrum code. The bug should be fixed in the next commits and you should be able to exercise the split-spectrum part of the code. However, it worth clarifying what the ESA folks mean with Hamming window. According to the "Sentinel-1 Level-1 Detailed Algorithm Definition" (https://sentinel.esa.int/documents/247904/1877131/Sentinel-1-Level-1-Detailed-Algorithm-Definition, page 9-14, paragraph 9-7), what is intended with "Hamming window" is instead a family of windows. Eq. 9-21 is what we have implemented in ISCE3 as "raised cosine window". I do not think we should add a new window in the split-spectrum module. We can direct the window type to "raised cosine" whenever we have a Hamming window in the S1 metadata, like is done here: https://github.com/vbrancat/COMPASS/blob/366ef67d240de755f0e514bb96a1e9f00b6d1790/src/compass/utils/range_split_spectrum.py#L88 . Does it sound good?

Reply:
Thanks you @vbrancat for the details. I took a look at the S1 spec you linked to and also I looked at the raised cosine in [isce3](https://github.com/isce-framework/isce3/blob/develop/python/packages/nisar/workflows/focus.py#L68). I'm not sure if the isce3 raised cosine is exactly same as the hamming in the s1 spec. I just tried to create the filter simply based on S1 spec as:

```
W = a-(1-a)*np.cos(np.linspace(0, 2*np.pi, num=nwin, endpoint=True))
```
I also got the filter from isce3. and I also just saw our isce2 implementation [here](https://github.com/isce-framework/isce2/blob/main/components/isceobj/TopsProc/runIon.py#L269). Here is a plot of the three implementation. I don't see that isce3 to be identical to what is in s1 spec. (Note that I shifted the isce2 filter by fftshift). The one labeled as s1 spec is the equation above.


<img width="576" alt="Screen Shot 2022-04-14 at 7 47 58 AM" src="https://user-images.githubusercontent.com/5033183/163417373-88bfe0d8-bd80-49ab-b32a-9504378d1ab3.png">

Reply:
Uhm, strange. I asked @jungkyoJung to check the Hamming window in ISCE3. I will submit a PR to include it, then. What do you mean you don't see any output? If you enabled split-spectrum, your geocoded burst should have 3 bands instead of 1. I tested it yesterday and that was the case.

Reply:
I was looking for three files. LOL. Now I see that you have put them in the same output file with three bands.

Reply:
Hi @vbrancat @hfattahi , 
The difference between windows in ISCE3 and the others come from how they are implemented. In ISCE3, the raised cosine was implemented as below. 
So, we have to use the adjusted `window_shape` variable to get identical shapes like `window_shape_hamming = 1 - 2* window_shape`. For example, if you used 0.5 for above plot, then we have to use `0` for cosine window so that we can get the identical windows. 
```
def cosine_window(n: int, pedestal: float):
    if not (0.0 <= pedestal <= 1.0):
        raise ValueError(f"Expected pedestal between 0 and 1, got {pedestal}.")
    b = 0.5 * (1 - pedestal)
    t = np.arange(n) - (n - 1) / 2.0
    return (1 - b) + b * np.cos(2 * np.pi * t / (n - 1))
```
![comparison_s1_isce3](https://user-images.githubusercontent.com/56169931/163470264-f338b27e-22b7-45b4-8050-cc34d3c117eb.png)

Reply:
Thanks @oberonia78 for the clarification. So that means we have to adjust the S1 coeffieceient before passing to isce3.

Reply:
Yes. Exactly.

Reply:
if window_shape is the `window_shape` coefficient coming from S1-A/B metadata, then the coefficient that should be passed to the `isce3` split-spectrum module is `1.0 - window_shape` and `window_type` being raised cosine? Correct @oberonia78 ?

Reply:
@vbrancat Your understanding is correct except `1.0 - window_shape`. This should be `1.0 - 2 * window_shape` for hamming.

Reply:
@oberonia78 but the new `window shape` would not be between 0 and 1 anymore, correct? This would cause the split spectrum code to crash. Say, if the window_shape we get from S1-A/B metadata is 0.75, the new window shape is -0.5.

Reply:
@vbrancat @hfattahi  Yes. You are right. I'll debug it.

Reply:
Ok to resolve this comment based on our discussion on isce3 repo, it looks like we should pass  "window_shape = 2 * burst.range_window_coefficient - 1", assuming S1 spec enforces the coefficient to be between 0.5 to 1.0 
@oberonia78 for completeness, would you please confirm and point to S1 documentation that the window coefficient is between 0.5 to 1.0 ?

Reply:
Thank you.